aiida_quantumespresso/cli/calculations/dos.py

```diff
@@ -0,0 +1,65 @@
+# -*- coding: utf-8 -*-
+"""Command line scripts to launch a `PwCalculation` for testing and demonstration purposes."""
+from __future__ import absolute_import
+import click
+
+from aiida.orm import RemoteData
+from aiida.cmdline.params import options, types
+from aiida.cmdline.utils import decorators
+
+from aiida_quantumespresso.cli.utils import options as options_qe
+
+# required inputs: code, input_parameters, (parent_calculation or parent_folder)
+# optional: settings
+
+
+@click.command()
+@options.CODE(required=True, type=types.CodeParamType(entry_point='quantumespresso.dos'))
+# Note: you cannot ask for a CalcJobNode created by a specific plugin, because you can get subclasses of Process via the
+# entry point system, but not subclasses of Node. You could filter by process_type of the CalcJobNode, but this is
+# currently not implemented in IdentifierParamType.
+#@options.CALCULATION(required=True, type=types.CalculationParamType(
+#    sub_classes=('aiida.calculations:quantumespresso.pw',)))
+@options.CALCULATION(required=True)
+@options_qe.MAX_NUM_MACHINES()
+@options_qe.MAX_WALLCLOCK_SECONDS()
+@options_qe.WITH_MPI()
+@options_qe.DAEMON()
+@decorators.with_dbenv()
+def cli(code, calculation, max_num_machines, max_wallclock_seconds, with_mpi, daemon):
+    """Run a DosCalculation."""
+    from aiida.engine import launch
+    from aiida.orm import Dict
+    from aiida.plugins import CalculationFactory
+    from aiida_quantumespresso.utils.resources import get_default_options
+
+    DosCalculation = CalculationFactory('quantumespresso.dos')  # pylint: disable=invalid-name
+
+    expected_process_type = 'aiida.calculations:quantumespresso.pw'
+    if calculation.process_type != expected_process_type:
+        raise click.BadParameter('The input calculation node has a process_type: {}; should be {}'.format(
+            calculation.process_type, expected_process_type))
+    parent_folder = calculation.get_outgoing(node_class=RemoteData, link_label_filter='remote_folder').one().node
+
+    inputs = {
+        'code': code,
+        'parent_folder': parent_folder,
+        'parameters': Dict(dict={}),
```

Comment:
If `settings` is optional, better not pass one if it is going to be empty. Same goes for `parameters` but I think your comment above said it *is* required. Which does not make sense if it is allowed to be empty. If that is true, then the `CalcJob` should make it not required

Reply:
Good points. I modeled the spec after `BasePwCpInputGenerator`, where `parameters` is required and `settings` is not. I'll make `parameters` and `settings` not required in `NamelistsCalculation` (which Dos inherits from), but I'm not sure if I should do the same for `BasePwCpInputGenerator`...

In general, is it better to add a `default=Dict(dict={})` or to just check for the presence of the input? The former is cleaner, but can add a useless empty input node.

Reply:
If the default is an empty dict, it is better not to specify it and just check for it in the code, assigning an empty dict there if not specified. This is because mutable defaults, such as dictionaries, are [dangerous](https://blog.florimond.dev/python-mutable-defaults-are-the-source-of-all-evil). The same story goes for AiiDA nodes and it has caused us a lot of problems in the past. We still use nodes as defaults for certain work chains, but we really shouldn't.

This is a more fundamental problem in `aiida-core` where we should forbid mutable defaults and only allow lambdas. They can then construct and store a default node each time it is invoked. This prevents the same node being passed around each time the process is launched. For the time being we should really limit using these defaults unless absolutely necessary.

Reply:
Thanks, very interesting.
I'll keep 'parameters' as required in `BasePwCpInputGenerator` for now, but make it (and 'settings') optional without a default in `NamelistsCalculation`.